src/QMCState.jl

```diff
@@ -0,0 +1,284 @@
+"""
+mutable struct ReplicaState{T,V,W,R<:FciqmcRunStrategy{T}}
+    v::V       # vector
+    w::W       # working memory. Maybe working memories could be shared among replicas?
+    pnorm::T   # previous walker number - used to control the shift
+    params::R  # params: step, laststep, dτ...
+    id::String # id is appended to column names
+end
+
+function ReplicaState(v, w, params, id="")
+    if isnothing(w)
+        w = similar(v)
+    end
+    return ReplicaState(v, w, walkernumber(v), params, id)
+end
+
+"""
+    QMCState
+
+Holds all inforamtion needed to run FCIQMC, except the data frame. Holds a `NTuple` of
+`ReplicaState`s and various strategies that control the algorithm.
+"""
+struct QMCState{
+    H,
+    N,
+    R<:ReplicaState,
+    MS<:MemoryStrategy,
+    RS<:ReportingStrategy,
+    SS<:ShiftStrategy,
+    TS<:TimeStepStrategy,
+    O,
+}
+    hamiltonian::H
+    replicas::NTuple{N,R}
+    maxlength::Int
+
+    m_strat::MS
+    r_strat::RS
+    s_strat::SS
+    τ_strat::TS
+
+    # Report x†⋅operator⋅y
+    # In the future, this should be replaced with a struct a la ReplicaStrategy.
+    operator::O
+end
+
+function QMCState(
+    hamiltonian, v;
+    laststep=nothing,
+    dτ=nothing,
+    threading=:auto,
+    wm=nothing,
+    params::FciqmcRunStrategy=RunTillLastStep{float(valtype(v))}(),
+    s_strat::ShiftStrategy=DoubleLogUpdate(),
+    r_strat::ReportingStrategy=EveryTimeStep(),
+    τ_strat::TimeStepStrategy=ConstantTimeStep(),
+    m_strat::MemoryStrategy=NoMemory(),
+    maxlength=2 * max(real(s_strat.targetwalkers), imag(s_strat.targetwalkers)),
+    num_replicas=1,
+    report_xHy=false,
+    operator=report_xHy ? hamiltonian : nothing,
+)
+    # Checks
+    if num_replicas < 1
+        error("need at least one replica.")
+    elseif !isnothing(operator) && num_replicas < 2
+        error(
+            "operator $operator needs at least two replicas. ",
+            "Set the number of replicas with the `num_replicas` keyowrd argument."
+        )
+    end
+    # Set up default arguments
+    r_strat = refine_r_strat(r_strat, hamiltonian)
+    if !isnothing(laststep)
+        params.laststep = laststep
+    end
+    if !isnothing(dτ)
+        params.dτ = dτ
+    end
+    wm = default_working_memory(threading, v, s_strat)
+    # Set up replicas
+    if num_replicas > 1
+        replicas = ntuple(num_replicas) do i
+            ReplicaState(deepcopy(v), deepcopy(wm), deepcopy(params), "_$i")
+        end
+    else
+        replicas = (ReplicaState(v, wm, params, ""),)
+    end
+
+    return QMCState(
+        hamiltonian, replicas, maxlength, m_strat, r_strat, s_strat, τ_strat, operator
+    )
+end
+
+function default_working_memory(threading, v, s_strat)
+    if threading == :auto
+        threading = max(real(s_strat.targetwalkers),imag(s_strat.targetwalkers)) ≥ 500
+    end
+    if threading && Threads.nthreads() > 1
+        return threadedWorkingMemory(v)
+    else
+        return similar(localpart(v))
```

Comment:
If `similar` does not conserve capacity, maybe it makes sense to do
```suggestion
        return sizehint!(similar(localpart(v), length(v))
```